addon/ember-concurrency.d.ts

```diff
@@ -0,0 +1,139 @@
+declare module 'ember-concurrency' {
+  import EmberObject from '@ember/object';
+  import {
+    UnwrapComputedPropertyGetter,
+    UnwrapComputedPropertyGetters
+  } from '@ember/object/-private/types';
+
+  import RSVP from 'rsvp';
+
+  // Lifted from @types/ember__object/observable.d.ts
+  interface Getter {
+    /**
+     * Retrieves the value of a property from the object.
+     */
+    get<K extends keyof this>(key: K): UnwrapComputedPropertyGetter<this[K]>;
+    /**
+     * To get the values of multiple properties at once, call `getProperties`
+     * with a list of strings or an array:
+     */
+    getProperties<K extends keyof this>(
+      list: K[]
+    ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+    getProperties<K extends keyof this>(
+      ...list: K[]
+    ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+  }
+
+  export type GeneratorFn<Args extends any[] = any[], R = any> = (
+    ...args: Args
+  ) => IterableIterator<R>;
+
+  export const all: typeof Promise.all;
+  export const allSettled: typeof RSVP.allSettled;
+  export const hash: typeof RSVP.hash;
+  export const race: typeof Promise.race;
+
+  export function timeout(ms: number): Promise<void>;
+
+  export function waitForEvent(
+    object: EmberObject | EventTarget,
+    eventName: string
+  ): Promise<Event>;
+
+  export function waitForProperty<T extends object, K extends keyof T>(
+    object: T,
+    key: K,
+    callbackOrValue: (value: T[K]) => boolean
+  ): Promise<void>;
+
+  export function waitForQueue(queueName: string): Promise<void>;
+
+  export function task<Args extends any[], R>(
+    taskFn: GeneratorFn<Args, R>
+  ): Task<Args, Exclude<R, Promise<any>>>;
```

Comment:
With TypeScript 3.6, this becomes much clearer / safer.

https://devblogs.microsoft.com/typescript/announcing-typescript-3-6-beta/#stricter-generators

Reply:
I'm *so* jazzed about that.

We should talk about TS version support at some point as well; it's probably fine to start with 3.6, given that and depending on when this lands and TS 3.6 lands, but the Typed Ember team is also working on hashing out a good ecosystem-wide versioning strategy as well to deal with the semi-regularity (and non-semver-compatible) breaking changes. Our current *actual* practice is "at least two most recent versions" but that's not going to work long-term.